### PR TITLE
Replace boost::units by LLNL/units

### LINKED ISSUE #1620: Find a better way to handle units
# What?

- `boost::units` may be too inflexible.
- Need unit scales.
- Need more than hard-coded set of supported unit combinations.

# Requirements

- Support units with different scales, e.g., `m` and `km`.
  - Note that powers of 10 are *not* sufficient, e.g., since we may have `s` and `h`.
  - A single scale for a given unit is also not sufficient, e.g., for `uAh` (microampere hours).
- Need support for adding custom units, e.g., for "counts".
- Need some way of interoperability with `boost::units` so we can keep making use of their physical constants, unless we want to replace that as well.
- Playing nicely with common Python packages for units would be an advantage.
- Do we want to support `unit = None`, rather than using `dimensionless` for, e.g., `dtype=string`?

# Considerations

- There appears to be very few C++ runtime unit libraries (but many compile-time libs):
  - https://github.com/LLNL/units seems most mature?
  - https://github.com/martinmoene/PhysUnits-RT 
- Python options may include `pint` and `astropy`
  - Would make a C++ library depend on a Python library?
  - Would introduce GIL into *all* of our operations!?
- Writting something from scratch is most likely far from trivial, contrary to what one might think at first glance.
- Is there a better way to make a runtime wrapper of `boost::units` that is less restrictive than what we have currently? JIT?
- Is it useful to limit what units a dtype can have? For example, as mask (bools) does not make sense with anything but `dimensionless`. Maybe more trouble than what is gained from being restrictive?

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Not ready yet, but at a point where we can play with it and see what is not working.

Known issues or things to note:
- `counts` is displayed as `item`. (worked around in our `to_string` wrapper)
- `meV` is displayed as `1.60217663399999985e-22J`. (fixed upstream)
- `gd_prtn_chrg` from a POWGEN file is displayed as `0.0035999999999999999C` (from `uAh` ?).  (fixed upstream)
- `counts/us` is translated `MBq`. OK?
- `counts/meV` shows as `6.2415090744607627e+21counts/J`
- Angstrom symbol uses different unicode character. (worked around in our `to_string` wrapper)
- Possible precision issue with `meV`, leading to test failures in some unit conversions. (turn out not to be the case, issue is different CODATA version).

Notes:
- The modified unit test for conversion via wavelength is because precise comparison failed. The reason turned out to be that boost uses `m_n` and `h` from CODATA 2006 whereas `LLNL/units` has CODATA 2018 (and so is the `m_n` I defined here, until we can get it added upstream).

Fixes #1620.

### CODE REVIEW

File: neutron/test/convert_test.cpp [hunk 1]
```diff
@@ -280,8 +280,9 @@ TEST_P(ConvertTest, Wavelength_to_Tof) {
   ASSERT_TRUE(tof.contains("counts"));
   // Broadcasting is needed as conversion introduces the dependance on
   // Dim::Spectrum
-  EXPECT_EQ(tof.coords()[Dim::Tof], broadcast(tof_original.coords()[Dim::Tof],
-                                              tof.coords()[Dim::Tof].dims()));
+  EXPECT_TRUE(all(is_approx(tof.coords()[Dim::Tof],
```
Comment (OwenArnold) [inline neutron/test/convert_test.cpp:283]: This is the precision loss problem you mentioned?

File: units/include/scipp/units/unit.h [hunk 2]
```diff
@@ -5,97 +5,19 @@
...
-  static constexpr Unit fromIndex(const int64_t index) {
-    Unit u;
-    u.m_index = index;
-    return u;
-  }
+  constexpr Unit(const llnl::units::precise_unit &u) noexcept : m_unit(u) {}
+  Unit(const std::string &unit);
 
-  constexpr scipp::index index() const noexcept { return m_index; }
+  constexpr auto underlying() const noexcept { return m_unit; }
```
Comment (OwenArnold) [inline units/include/scipp/units/unit.h:20]: Could this be made private and operator level support provided through `Unit` directly, such as `operator * (...)` This would better encapsulate the use of the underlying unit library/type

File: units/unit.cpp [hunk 3]
```diff
@@ -3,79 +3,47 @@
...
+Unit::Unit(const std::string &unit)
+    : Unit(llnl::units::unit_from_string(unit == "dimensionless" ? "" : unit)) {
 }
 
 std::string Unit::name() const {
-  static const auto names = make_name_lut(supported_units_t{});
-  return names[index()];
+  auto repr = to_string(m_unit);
+  // Replace 'LATIN CAPITAL LETTER A WITH RING ABOVE' by 'ANGSTROM SIGN'
+  repr = std::regex_replace(repr, std::regex("Å"), "Å");
```
Comment (OwenArnold) [inline units/unit.cpp:23]: Search and replacement terms the same?

File: units/unit.cpp [hunk 4]
```diff
@@ -3,79 +3,47 @@
...
+  repr = std::regex_replace(repr, std::regex("^u"), "µ");
+  repr = std::regex_replace(repr, std::regex("item"), "counts");
+  return repr == "" ? "dimensionless" : repr;
 }
```
Comment (OwenArnold) [inline units/unit.cpp:27]: do we need to worry about the inverse operation of this i.e. creating Unit from name str. If so, maybe best to define a common map (or vector of pairs) that can be used both ways?

File: variable/string.cpp [hunk 5]
```diff
@@ -3,6 +3,7 @@
 /// @file
 /// @author Simon Heybrock
 #include <chrono>
+#include <iomanip>
```
Comment (OwenArnold) [inline variable/string.cpp:6]: Why is this needed?

**Review by OwenArnold — COMMENTED** [review]
Not tried out in practice yet, but looks good thus far.

File: neutron/test/convert_test.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply neutron/test/convert_test.cpp:283]: Well, it was what I mentioned, but it turned out that it is not due to precision loss, but different (old) CODATA used in boost. By chance there seemed to be an exact double match before, but not now.

File: variable/string.cpp (hunk 5 above)
Comment (SimonHeybrock) [reply variable/string.cpp:6]: Because I removed some std headers from `unit.h` and this did not compile any more.

File: units/unit.cpp (hunk 4 above)
Comment (SimonHeybrock) [reply units/unit.cpp:27]: Reverse seems to work. But my goal is to make things configureable upstream, so string reprs of certain units can be defined by the lib user. For example, they use `hr` for formatting hours, but I would like to have `h` (SI).

File: neutron/test/convert_test.cpp (hunk 1 above)
Comment (OwenArnold) [reply neutron/test/convert_test.cpp:283]: I see you have actually been able to tighten tolerances elsewhere.

File: units/unit.cpp (hunk 3 above)
Comment (OwenArnold) [reply units/unit.cpp:23]: I thought it might be that. Should have gone to raw view.

**Comment by nvaytet** [comment]
I'm trying to build this branch to play with it but getting these errors:
```
[  4%] Building CXX object units/CMakeFiles/scipp-units.dir/cmake_pch.hxx.gch
In file included from /home/nvaytet/code/scipp/source/units/include/scipp/units/string.h:12,
                 from /home/nvaytet/code/scipp/source/units/include/scipp/units/except.h:9,
                 from /home/nvaytet/code/scipp/build/units/CMakeFiles/scipp-units.dir/cmake_pch.hxx:5,
                 from <command-line>:
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:17:24: error: ‘llnl’ does not name a type
   17 |   constexpr Unit(const llnl::units::precise_unit &u) noexcept : m_unit(u) {}
      |                        ^~~~
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:17:50: error: expected unqualified-id before ‘&’ token
   17 |   constexpr Unit(const llnl::units::precise_unit &u) noexcept : m_unit(u) {}
      |                                                  ^
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:17:49: error: expected ‘)’ before ‘&’ token
   17 |   constexpr Unit(const llnl::units::precise_unit &u) noexcept : m_unit(u) {}
      |                 ~                               ^~
      |                                                 )
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:17:50: error: constructors may not be ref-qualified
   17 |   constexpr Unit(const llnl::units::precise_unit &u) noexcept : m_unit(u) {}
      |                                                  ^
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:17:50: error: expected ‘;’ at end of member declaration
   17 |   constexpr Unit(const llnl::units::precise_unit &u) noexcept : m_unit(u) {}
      |                                                  ^
      |                                                   ;
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:17:51: error: ‘u’ does not name a type
   17 |   constexpr Unit(const llnl::units::precise_unit &u) noexcept : m_unit(u) {}
      |                                                   ^
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:36:3: error: ‘llnl’ does not name a type
   36 |   llnl::units::precise_unit m_unit;
      |   ^~~~
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h: In member function ‘constexpr auto scipp::units::Unit::underlying() const’:
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:20:55: error: ‘m_unit’ was not declared in this scope
   20 |   constexpr auto underlying() const noexcept { return m_unit; }
      |                                                       ^~~~~~
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h: At global scope:
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:55:30: error: ‘llnl’ has not been declared
   55 | constexpr Unit dimensionless{llnl::units::precise::one};
      |                              ^~~~
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:56:20: error: ‘llnl’ has not been declared
   56 | constexpr Unit one{llnl::units::precise::one}; /// alias for dimensionless
      |                    ^~~~
/home/nvaytet/code/scipp/source/units/include/scipp/units/unit.h:57:18: error: ‘llnl’ has not been declared
   57 | constexpr Unit m{llnl::units::precise::meter};
```

File: units/unit.cpp (hunk 3 above)
Comment (jl-wynen) [reply units/unit.cpp:23]: But both are `'ANGSTROM SIGN'`, no?

Comment (SimonHeybrock) [reply units/unit.cpp:23]: No, see comment in the line above. The replaced one is https://www.compart.com/en/unicode/U+00C5.

Comment (jl-wynen) [reply units/unit.cpp:23]: Ah, ok

**Comment by SimonHeybrock** [comment]
@nvaytet Did cmake run successfully, getting the library? Have you tried this before (last week)? The upstream has changed to my fork, try deleting all `llnl*` folders in your build folder, maybe something is messed up there?

**Comment by jl-wynen** [comment]
Getting the same error with a clean build

**Comment by SimonHeybrock** [comment]
@nvaytet @jl-wynen I pushed an update, try again (have to escape `:` in cmake).

**Comment by jl-wynen** [comment]
Still getting the error

**Comment by SimonHeybrock** [comment]
> Still getting the error

I love cmake... clean build fails, then I change the string to include or exclude escaped characters for `:`, then it passes a non-clean build.

**Comment by nvaytet** [comment]
So far, my experience with this is pretty good. Tried it in the imaging notebook and it's reduced the number of places where i've had to use `.values` and then re-add units later.

One thing that was a bit weird is that it does seem to perform some automatic conversions.
If I divide meters by microseconds, I get `km/ms` and not `m/us`. I guess it's still valid but a little unintuitive.

**Comment by SimonHeybrock** [comment]
As discussed on Slack: Merging, and further issues will be addressed later.
